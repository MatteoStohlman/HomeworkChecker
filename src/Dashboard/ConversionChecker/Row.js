import React from 'react';
import {compose,withState,withHandlers} from 'recompose';
import { withStyles } from '@material-ui/core/styles';
import TableCell from '@material-ui/core/TableCell';
import TableRow from '@material-ui/core/TableRow';
import UnitPicker from 'Components/UnitPicker'
import TextField from '@material-ui/core/TextField';
import math from 'mathjs'
import DeleteIcon from 'Components/DeleteIcon'

const styles= {
  row:{
    backgroundColor:props=>props.studentError?'rgb(255, 224, 224)':props.studentSuccess?'rgb(223, 239, 216)':'white',
  }
}

const ConversionRow = ({
  //STATE
    fromValue,handleFromUnitSelect,handleFromValueChange,
    toType,toUnit,handleToUnitSelect,
    studentResponse,handleStudentResponseChange,
    studentError,setStudentError,
  //HANDLERS
    calculateResult,handleDelete,
  //OTHER
    classes,...props
})=> {
  let result = calculateResult();
  return (
    <TableRow className={classes.row} data-cy='row'>
      <TableCell>
        <TextField
          className='fromInput'
          value={fromValue}
          onChange={(e)=>handleFromValueChange(e.target.value)}
          type="number"
          InputProps={{disableUnderline:true,"data-cy":'fromValue'}}
          placeholder='value'
          InputLabelProps={{
            shrink: true,
          }}
        />
      </TableCell>
      <TableCell>
        <UnitPicker inputProps={{"data-cy":'fromUnit'}} onChange={handleFromUnitSelect}/>
      </TableCell>
      <TableCell>
        <UnitPicker inputProps={{"data-cy":'toUnit'}} units={toType} value={toUnit} onChange={handleToUnitSelect}/>
      </TableCell>
      <TableCell>
        <TextField
          className='toInput'
          value={studentResponse}
          onChange={(e)=>handleStudentResponseChange(e.target.value)}
          type="number"
          InputProps={{disableUnderline:true,"data-cy":'guess'}}
          placeholder='value'
          InputLabelProps={{
            shrink: true,
          }}
        />
      </TableCell>
      <TableCell>
        <div data-cy='result'>
          {result}
        </div>
      </TableCell>
      <TableCell>
        <DeleteIcon onClick={handleDelete}/>
      </TableCell>
    </TableRow>
  )
}

export default compose(
  withState('fromValue','setFromValue',""),
  withState('fromUnit','setFromUnit',null),
  withState('fromType','setFromType',null),
  withState('toValue','setToValue',null),
  withState('toUnit','setToUnit',null),
  withState('toType','setToType',null),
  withState('studentResponse','setStudentResponse',""),
  withState('studentError','setStudentError',false),
  withState('studentSuccess','setStudentSuccess',false),
  withHandlers({
    //Compare conversion "TO" value to Student Response value and set state
    calculateError:props=>(toValue)=>{
      let shouldShowError = false;
      let shouldShowSuccess = false;
      if(toValue&&props.studentResponse){
        let roundedToValue = math.round(toValue,1)
        let roundedResp = math.round(props.studentResponse,1)
        if(roundedToValue!==roundedResp){
          shouldShowError=true
        }else{
          shouldShowSuccess=true
        }
      }
      if(props.studentError!==shouldShowError){
        props.setStudentError(shouldShowError)
      }
      if(props.studentSuccess!==shouldShowSuccess){
        props.setStudentSuccess(shouldShowSuccess)
      }
    }
  }),
  withHandlers({
    //Calculate Conversion "TO" value
    calculateResult:props=>(overrides)=>{
      let {fromValue,fromUnit,toUnit} = props
      if(overrides){
        fromValue=overrides.fromValue||fromValue
        fromUnit=overrides.fromUnit||fromUnit
        toUnit=overrides.toUnit||toUnit
      }
      if(fromValue&&fromUnit&&toUnit){
        let from = math.unit(fromValue,fromUnit.toLowerCase())
        let toValue = from.toNumber(toUnit.toLowerCase())
        props.calculateError(toValue)
        return toValue&&math.round(toValue,1)
      }else{
        return false
      }
    }
  }),
  withHandlers({
    handleFromUnitSelect:props=>(selectedUnit)=>{
      if(!selectedUnit){//If the selected value is null e.g. the input is cleared
        props.setFromType(null)
        props.setFromUnit(null)
        return;
      }
      props.setFromType(selectedUnit.type)
      props.setFromUnit(selectedUnit.value)
      if(selectedUnit.type!==props.toType){//From and to Unit types must match. Cannot convert temp to volume
        props.setToType(selectedUnit.type)
        props.setToUnit(null)
      }
    },
    handleToUnitSelect:props=>(selectedUnit)=>{
      if(!selectedUnit){//If the selected value is null e.g. the input is cleared
        props.setToType(null)
        props.setToUnit(null)
        return
      }
      props.setToType(selectedUnit.type)
      props.setToUnit(selectedUnit.value)
    },
    handleFromValueChange:props=>(value)=>{
      props.setFromValue(value)
    },
    handleStudentResponseChange:props=>(value)=>{
      props.setStudentResponse(value)
    },
    handleDelete:props=>()=>{
      props.onDelete&&props.onDelete()
    }
  }),
  withStyles(styles)
)(ConversionRow)
