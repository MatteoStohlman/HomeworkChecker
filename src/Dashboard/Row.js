import React from 'react';
import Typography from '@material-ui/core/Typography';
import {compose,withProps,withState,withHandlers} from 'recompose';
import { withStyles } from '@material-ui/core/styles';
import TableCell from '@material-ui/core/TableCell';
import TableRow from '@material-ui/core/TableRow';
import UnitPicker from 'Components/UnitPicker'
import TextField from '@material-ui/core/TextField';
import math from 'mathjs'
const COMPONENT_NAME = ({
  //PROPS
    //required

    //optional
      variant='body1',
    //calculated

  //STATE
    fromValue,handleFromUnitSelect,handleFromValueChange,
    toType,toUnit,handleToUnitSelect,
    studentResponse,handleStudentResponseChange,
    studentError,setStudentError,
  //HANDLERS
    calculateResult,
  //OTHER
    ...props
})=> {
  return (
    <TableRow>
      <TableCell>
        <TextField
          value={fromValue}
          onChange={(e)=>handleFromValueChange(e.target.value)}
          type="number"
          InputLabelProps={{
            shrink: true,
          }}
        />
      </TableCell>
      <TableCell>
        <UnitPicker onChange={handleFromUnitSelect}/>
      </TableCell>
      <TableCell>
        <UnitPicker units={toType} value={toUnit} onChange={handleToUnitSelect}/>
      </TableCell>
      <TableCell>
        <TextField
          error={studentError}
          value={studentResponse}
          onChange={(e)=>handleStudentResponseChange(e.target.value)}
          type="number"
          InputLabelProps={{
            shrink: true,
          }}
        />
      </TableCell>
      <TableCell>
        {calculateResult()}
      </TableCell>
    </TableRow>
  )
}

export default compose(
  withState('fromValue','setFromValue',null),
  withState('fromUnit','setFromUnit',null),
  withState('fromType','setFromType',null),
  withState('toValue','setToValue',null),
  withState('toUnit','setToUnit',null),
  withState('toType','setToType',null),
  withState('studentResponse','setStudentResponse',null),
  withState('studentError','setStudentError',false),
  withHandlers({
    calculateError:props=>(toValue)=>{
      let shouldShowError = false;
      if(toValue&&props.studentResponse){
        let roundedToValue = math.round(toValue,1)
        let roundedResp = math.round(props.studentResponse,1)
        if(roundedToValue!=roundedResp){
          shouldShowError=true
        }
      }
      if(props.studentError!==shouldShowError){
        props.setStudentError(shouldShowError)
      }
    }
  }),
  withHandlers({
    calculateResult:props=>(overrides)=>{
      let {fromValue,fromUnit,toUnit} = props
      if(overrides){
        fromValue=overrides.fromValue||fromValue
        fromUnit=overrides.fromUnit||fromUnit
        toUnit=overrides.toUnit||toUnit
      }
      if(fromValue&&fromUnit&&toUnit){
        let from = math.unit(fromValue,fromUnit.toLowerCase())
        let toValue = from.toNumber(toUnit.toLowerCase())
        props.calculateError(toValue)
        return toValue&&math.round(toValue,1)
      }else{
        console.log('Invalid values for Calculation');
        return false
      }
    }
  }),
  withHandlers({
    handleFromUnitSelect:props=>(selectedUnit)=>{
      if(!selectedUnit){//If the selected value is null e.g. the input is cleared
        props.setFromType(null)
        props.setFromUnit(null)
        return;
      }
      props.setFromType(selectedUnit.type)
      props.setFromUnit(selectedUnit.value)
      if(selectedUnit.type!=props.toType){//From and to Unit types must match. Cannot convert temp to volume
        props.setToType(selectedUnit.type)
        props.setToUnit(null)
      }
    },
    handleToUnitSelect:props=>(selectedUnit)=>{
      if(!selectedUnit){//If the selected value is null e.g. the input is cleared
        props.setToType(null)
        props.setToUnit(null)
        return
      }
      props.setToType(selectedUnit.type)
      props.setToUnit(selectedUnit.value)
    },
    handleFromValueChange:props=>(value)=>{
      props.setFromValue(value)
    },
    handleStudentResponseChange:props=>(value)=>{
      props.setStudentResponse(value)
    }
  })
)(COMPONENT_NAME)
