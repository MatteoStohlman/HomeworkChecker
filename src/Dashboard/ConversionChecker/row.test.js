import React from 'react';
import { mount , shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import Row from './row';

describe('ConversionTableRow',() => {
  test('basic Row renders and snapshot', () => {
    shallow(<Row/>)
    const picker = renderer.create(
      <Row/>
    );
    expect(picker.toJSON()).toMatchSnapshot();
  });

  test('row calculates to value',() => {
    const row = mount(<table><tbody><Row/></tbody></table>)
    row.find('.fromInput').first().find('input').first().props().onChange({target:{value:'32'}})
    row.find('UnitPicker').first().props().onChange({type:'temperature',value:'fahrenheit'})
    row.find('UnitPicker').at(1).props().onChange({type:'temperature',value:'celsius'})
    row.update()
    console.log(row.debug());
  })
})
