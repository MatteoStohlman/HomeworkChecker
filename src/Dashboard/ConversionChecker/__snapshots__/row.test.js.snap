// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionTableRow basic Row renders and snapshot 1`] = `
<tr
  className="MuiTableRow-root ConversionRow-row-1 ConversionRow-row-2"
>
  <td
    aria-sort={null}
    className="MuiTableCell-root"
  >
    <div
      className="MuiFormControl-root MuiTextField-root fromInput"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="value"
          required={false}
          type="number"
          value=""
        />
      </div>
    </div>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root"
  >
    <div
      className=" css-2b097c-container"
      onKeyDown={[Function]}
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <div
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input UnitPicker-input-53"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            required={false}
            type="text"
          >
            <div
              className="UnitPicker-valueContainer-54"
            >
              <p
                className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary UnitPicker-placeholder-57"
              >
                pick units
              </p>
              <div
                className="css-1g6gooi"
              >
                <div
                  className=""
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <input
                    aria-autocomplete="list"
                    autoCapitalize="none"
                    autoComplete="off"
                    autoCorrect="off"
                    disabled={false}
                    id="react-select-2-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    spellCheck="false"
                    style={
                      Object {
                        "background": 0,
                        "border": 0,
                        "boxSizing": "content-box",
                        "color": "inherit",
                        "fontSize": "inherit",
                        "label": "input",
                        "opacity": 1,
                        "outline": 0,
                        "padding": 0,
                        "width": "1px",
                      }
                    }
                    tabIndex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style={
                      Object {
                        "height": 0,
                        "left": 0,
                        "overflow": "scroll",
                        "position": "absolute",
                        "top": 0,
                        "visibility": "hidden",
                        "whiteSpace": "pre",
                      }
                    }
                  >
                    
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root"
  >
    <div
      className=" css-2b097c-container"
      onKeyDown={[Function]}
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <div
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input UnitPicker-input-53"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            required={false}
            type="text"
          >
            <div
              className="UnitPicker-valueContainer-54"
            >
              <p
                className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary UnitPicker-placeholder-57"
              >
                pick units
              </p>
              <div
                className="css-1g6gooi"
              >
                <div
                  className=""
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <input
                    aria-autocomplete="list"
                    autoCapitalize="none"
                    autoComplete="off"
                    autoCorrect="off"
                    disabled={false}
                    id="react-select-3-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    spellCheck="false"
                    style={
                      Object {
                        "background": 0,
                        "border": 0,
                        "boxSizing": "content-box",
                        "color": "inherit",
                        "fontSize": "inherit",
                        "label": "input",
                        "opacity": 1,
                        "outline": 0,
                        "padding": 0,
                        "width": "1px",
                      }
                    }
                    tabIndex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style={
                      Object {
                        "height": 0,
                        "left": 0,
                        "overflow": "scroll",
                        "position": "absolute",
                        "top": 0,
                        "visibility": "hidden",
                        "whiteSpace": "pre",
                      }
                    }
                  >
                    
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root"
  >
    <div
      className="MuiFormControl-root MuiTextField-root toInput"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="value"
          required={false}
          type="number"
          value=""
        />
      </div>
    </div>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root"
  />
  <td
    aria-sort={null}
    className="MuiTableCell-root"
  >
    <span
      aria-hidden={true}
      className="material-icons MuiIcon-root DeleteIcon-root-90"
      onClick={[Function]}
    >
      delete
    </span>
  </td>
</tr>
`;
